### PR TITLE
Bug/607 fidelity statevector kernel cannot be pickled

### LINKED ISSUE #607: FidelityStatevectorKernel cannot be pickled
### Environment

- **Qiskit Machine Learning version**: qiskit-machine-learning==0.6.0
- **Python version**:  Python 3.10.4
- **Operating system**:  Ubuntu 20.04.5 LTS

### What is happening?

qiskit_machine_learning.kernels.fidelity_statevector_kernel.FidelityStatevectorKernel cannot be pickled.  I.e., it gives error:

> _pickle.PicklingError: Can't pickle <functools._lru_cache_wrapper object at 0x7f7de13c7e20>: it's not the same object as qiskit_machine_learning.kernels.fidelity_statevector_kernel.FidelityStatevectorKernel._get_statevector_

Being able to pickle objects is required for a lot of frameworks we'd want to use machine learning components with.  
I.e., this error prevents a lot of things - such as using it in a parallelization framework like ray (which requires pickling) or using it with common machine learning approaches such as sklearn GridSearchCV which also requires pickling.

The cause is a known issue of using lru_cache as is done in the code for FidelityStatevectorKernel (line 100-101):
```
# Create the statevector cache at the instance level.
self._get_statevector = lru_cache(maxsize=cache_size)(self._get_statevector_)
```

E.g., see: https://stackoverflow.com/questions/54363043/pickling-lru-cached-function-on-object
There are a number of fixes, including some listed in the SO post above.

### How can we reproduce the issue?

```
from qiskit.circuit.library import PauliFeatureMap

from qiskit_machine_learning.kernels.fidelity_statevector_kernel import (
    FidelityStatevectorKernel,
)

import pickle


kernel = FidelityStatevectorKernel(
    feature_map=PauliFeatureMap(
        feature_dimension=3,
        reps=1,
        entanglement="linear",
        alpha=1.0,
        paulis=("Z", "ZZ"),
    )
)

kernel_obj = pickle.dumps(kernel)
```

### What should happen?

Fix the class so it can be pickled.  I.e., the above code should not throw an error, but allow successfully pickling the object.

### Any suggestions?

You can look at the stack overflow post linked above, and other similar posts.  There are some easy fixes, such as writing __getstate__ and __setstate__ to handle that attribute specially.

### PR DESCRIPTION
**Pull request by oscar-wallis**
### Summary
Fixes #607 where FidelityStatevectorKernel was throwing an error when pickled due to a known pickling error with functools.lru_cache. This has been fixed by adding a self.cache_size attribute and custom __getstate__ and __setstate__ functions used when pickling and unpickling.

### Details and comments
When pickled the lru_cache is set to None in the object dict. When unpicked the cache is re-initiliased with the new cache_size variable and __getstate__ and __setstate__. Unit tests added to check object pickling and unpickling as well as manually using __getstate__ and __setstate__ to initialise a FidelityStatevectorKernel. Passed all other unit tests.

### CODE REVIEW

File: qiskit_machine_learning/kernels/fidelity_statevector_kernel.py [hunk 1]
```diff
@@ -97,7 +97,7 @@ def __init__(
         self._auto_clear_cache = auto_clear_cache
         self._shots = shots
         self._enforce_psd = enforce_psd
-
+        self.cache_size = cache_size
```
Comment (woodsp-ibm) [inline qiskit_machine_learning/kernels/fidelity_statevector_kernel.py]: Should this be `self._cache` ie something we consider private and uses should not change. At present its a public instance var and part of the public API - something we document via Attributes in this case - but in this case its not something I think we want users to change other that via the constructor.

File: releasenotes/notes/fix-fid_statevector_kernel-pickling-b7fa2b13a15ec9c6.yaml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+---
+fixes:
+  - |
+    Fixed a bug where :class:`~qiskit_machine_learning.kernels.FidelityQuantumKernel` threw an error when pickled. This was due
```
Comment (woodsp-ibm) [inline releasenotes/notes/fix-fid_statevector_kernel-pickling-b7fa2b13a15ec9c6.yaml:4]: Just as a note you can do this too, as long as its a class that is unique ''':class:`.FidelityQuantumKernel` ``` where with the `.` sphinx will find the class without you having to type the full thing. 

Unless any of how it was fixed affects the end user then just stating its fixed and can be pickled as expected now suffices. The release notes are really just the public API and changes/fixes perceived when using the API for end-users - internal details of how it was fixed, unless they have additional impacts in some way are not really relevant to such users. Eg. to fix things we could have stopped user lru_cache and done something else internally. All the end user needs to know is that the problem is fixed. Hence to me only the content of the first sentence is needed here for a release note for this.

**Review by woodsp-ibm — COMMENTED** [review]
Normally if we have a bug that does not show up in the test cases it suggests we need one to test that aspects for better coverage and to make sure it things stay working. I think a test case that pickles and unpickles is needed.

File: qiskit_machine_learning/kernels/fidelity_statevector_kernel.py (hunk 1 above)
Comment (oscar-wallis) [reply qiskit_machine_learning/kernels/fidelity_statevector_kernel.py]: yes it should be _cache_size, changed now

File: releasenotes/notes/fix-fid_statevector_kernel-pickling-b7fa2b13a15ec9c6.yaml (hunk 2 above)
Comment (oscar-wallis) [reply releasenotes/notes/fix-fid_statevector_kernel-pickling-b7fa2b13a15ec9c6.yaml:4]: Ah, okay I will reduce it down to the first sentence.

**Comment by oscar-wallis** [comment]
> Normally if we have a bug that does not show up in the test cases it suggests we need one to test that aspects for better coverage and to make sure it things stay working. I think a test case that pickles and unpickles is needed.

Does that mean a test is needed where we pickle and unpickle all qml objects?

**Comment by woodsp-ibm** [comment]
> Does that mean a test is needed where we pickle and unpickle all qml objects?

Seems like a lot of objects! This is much more an edge case I guess due to lru_cache so I would leave it at just a test for this object - makes sure it works in CI over the platforms and continues to do so.

**Comment by oscar-wallis** [comment]
Added unit testing and edited release notes.

File: test/kernels/test_fidelity_statevector_kernel.py [hunk 3]
```diff
@@ -343,6 +344,52 @@ def test_properties(self):
...
+            self.assertNotEqual(kernel1, kernel2)
+
+        with self.subTest("Fail if objects 1 and 3 are not unique"):
+            self.assertNotEqual(kernel1, kernel3)
+
+        with self.subTest("Fail if caches 1 and 2 are incorrectly initialised"):
+            self.assertEqual(type(kernel1._get_statevector), type(kernel2._get_statevector))
+
+        with self.subTest("Fail if caches 1 and 3 are incorrectly initialised"):
+            self.assertEqual(type(kernel1._get_statevector), type(kernel3._get_statevector))
```
Comment (OkuyanBoga) [inline test/kernels/test_fidelity_statevector_kernel.py:375]: To make testing clearer, can you please test messages more informative?

Comment (OkuyanBoga) [reply test/kernels/test_fidelity_statevector_kernel.py:375]: When the tests fail, it would be hard to understand for the people, for example, where it is originated
Instead of `"Fail if objects 1 and 3 are not the same type"`, maybe, something like 
`"Pickle test failed, kernels are not the same type"`.

Comment (oscar-wallis) [reply test/kernels/test_fidelity_statevector_kernel.py]: Are these better 83a00d3 ? B)

Comment (OkuyanBoga) [reply test/kernels/test_fidelity_statevector_kernel.py]: Sounds better, thanks

File: test/kernels/test_fidelity_statevector_kernel.py [hunk 4]
```diff
@@ -343,6 +344,55 @@ def test_properties(self):
...
+        with self.subTest("Pickle fail, kernels are not unique objects"):
+            self.assertNotEqual(kernel1, kernel2)
+
+        with self.subTest("Pickle fail, kernels are not unique objects"):
+            self.assertNotEqual(kernel1, kernel3)
+
+        with self.subTest("Pickle fail, caches are not the same type"):
+            self.assertEqual(type(kernel1._get_statevector), type(kernel2._get_statevector))
+
+        with self.subTest("Pickel fail, caches are not the same type"):
```
Comment (declanmillar) [inline test/kernels/test_fidelity_statevector_kernel.py:375]:
```suggestion
        with self.subTest("Pickle fail, caches are not the same type"):
```

**Review by declanmillar — CHANGES_REQUESTED** [review]
Thanks very much for addressing this issue! I just have a few minor suggestions.

File: .gitignore [hunk 5]
```diff
@@ -127,3 +127,4 @@ venv.bak/
 
 # stestr repostiory
 .stestr/
+bug607_replicated.py
```
[the remark this answers is not in the document]
Comment (oscar-wallis) [reply .gitignore]: Yes, it was added accidentally. I just wanted to make sure I didn't upload the bug replicating file, I didn't realise the gitignore is added here too.

**Comment by oscar-wallis** [comment]
> Thanks very much for addressing this issue! I just have a few minor suggestions.

I appreciate the suggestions! I'm surprised some of them weren't picked up by CI. All the changes were approved.

**Comment by declanmillar** [comment]
You'll need to add `from typing import Any` to `fidelity_statevector_kernel.py`, but, unfortunately, I'm not allowed to make that suggestion as it's in an unedited code block.

**Comment by oscar-wallis** [comment]
> `from typing import Any`

Can do, why does it need to be added?

**Review by adekusar-drl — APPROVED** [review]
Looks good to me. Thank you.

**Review by edoaltamura — APPROVED** [review]
